### PR TITLE
Implemented Segment Tree Data Structure

### PR DESCRIPTION
**Pull request by Ramy-Badr-Ahmed**
**Contents:**

- `SegmentTreeNode` Class: Represents individual nodes in the Segment Tree with attributes `start`, `end`, `value`, `left`, and `right`. Where:

    - `start` and `end`: Define the range of the array segment covered by this node.
     - `value`: Stores the aggregated result (e.g., sum, min, max) for this segment, determined by the callback function passed during construction. Defaults to arithmetic sum if no callback is provided.
    - `left` and `right`: Pointers to the left and right child nodes, allowing traversal through the tree.

     The `SegmentTreeNode` class encapsulates the ranges and results, linking to left and right child nodes.

- `SegmentTree` Class: Implements the core functionalities of a Segment Tree.

     - `buildTree`: Constructs the tree from the input array, using a callback function to define custom operations (e.g., sum, max).
     - `update`: Updates the value at a specified index, adjusting the tree accordingly with the `updateTree` method.
     - `rangeUpdate`: Updates values across a specified range, propagating changes through the tree with the `rangeUpdateTree` method.
     - `query`: Retrieves an aggregated result (e.g., sum, max) from a specified array range using the `queryTree` method for traversal.
     - `getCurrentArray`: Returns the current array representation of the segment tree.
     - `getRoot`: Returns the root node of the segment tree.
     - `serialize`: Converts the segment tree into a JSON string.
     - `deserialize`: Restores it from the serialized format.

     The class supports arbitrary operations (e.g., sum, min, max) using callback functions, or falls back to a sum-based tree if no callback is provided.

- Unit Tests:

     `SegmentTreeTest.php`: Contains PHPUnit tests to validate the correct behavior of segment tree building, querying, updating, performing range updates, and serialization and deserialization to ensure persistence of tree structure.

**Time Complexity**

A time complexity of `O(log n)` for both `query` and `update` operations, where `n` is the size of the input array. The build operation takes `O(n log n)` in total.

**GitHub Actions**

All tests and workflows (in my forked repository) have passed:

- [PHPCS](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/10981204979/job/30487953342)

[Code style](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/workflows/code-style.yml)

- [DIRECTORY.md](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/10981204985/job/30487953304)
 
[directory_md](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/workflows/directory_workflow.yml)

- [build](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/runs/10981204981/job/30487953359)
 
[PHP Composer](https://github.com/Ramy-Badr-Ahmed/PHP-DSA/actions/workflows/ci.yml)

### CODE REVIEW

File: composer.json
Comment (Ramy-Badr-Ahmed) [inline composer.json:1]: Added to composer.json for the `serialize` and `deserialize` methods which use `json_encode` and `json_decode`.

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
Hi @darwinz 
Looking forward to your review 🙂

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
Added Doc blocks to the rest of my files requesting mentioning me for bugs/corrections.

**Review by darwinz — APPROVED** [review]
Thank you for the addition!

**Comment by Ramy-Badr-Ahmed** [comment]
> Thank you for the addition!

Thanks for the review and merge  🙂

**Pull request merged**